### PR TITLE
Feature/table

### PR DESCRIPTION
**Pull request by frozenhelium**

### CODE REVIEW

**Comment by AdityaKhatri** [comment]
Remove predefined columns as suggested by @tnagorra .
Also, fix typecheck issues.

File: src/stories/Table/Table.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    const filterState = useFilterState();
+    const { filtering } = filterState;
+
+    const orderState = useOrderState(staticColumnOrdering);
+    const { ordering } = orderState;
+
+    const orderedColumns = useOrdering(columns, ordering);
+    const filteredData = useFiltering(filtering, orderedColumns, data);
+    const sortedData = useSorting(sorting, orderedColumns, filteredData);
```
Comment (tnagorra) [inline src/stories/Table/Table.stories.tsx:121]: We should also comment this out!

File: src/stories/Table/Table.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+];
+
+const staticColumnOrdering = [
+    { name: 'id' },
+    { name: 'name' },
+    { name: 'budget' },
+    { name: 'date' },
+    { name: 'datetime' },
+    { name: 'aboveBudget' },
+];
```
Comment (tnagorra) [inline src/stories/Table/Table.stories.tsx:105]: If we removed the columns 'date', 'dateTime', 'aboveBudget' then we have to remove these from staticColumnOrdering

File: src/index.tsx [hunk 3]
```diff
@@ -66,6 +66,8 @@ export { default as SelectInput } from './components/SelectInput';
 export type { Props as SelectInputProps } from './components/SelectInput';
 export type { Props as TabProps } from './components/Tab';
 export { default as Tab } from './components/Tab';
+export type { Props as TableProps } from './components/Table';
+export { default as Table } from './components/Table';
```
Comment (tnagorra) [inline src/index.tsx]: We need to export more than the Table component

File: src/hooks/useDropHandler.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+
+interface DragHandler<T> {
+    (e: React.DragEvent<T>): void;
+}
+
+function useDropHandler(
```
Comment (tnagorra) [inline src/hooks/useDropHandler.tsx:7]: Please use this in DropContainer

File: src/components/Table/predefinedColumns.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        headerCellRendererParams: {
+            sortable: options?.sortable,
+            filterType: options?.filterType,
+            orderable: options?.orderable,
+            hideable: options?.hideable,
+        },
+        cellRenderer: NumberOutput,
+        cellRendererParams: (_: K, datum: D): NumberOutputProps => ({
+            value: accessor(datum),
+            placeholder: 'N/a',
```
Comment (tnagorra) [inline src/components/Table/predefinedColumns.ts]:
```suggestion
```

File: src/stories/Table/Table.stories.tsx (hunk 1 above)
Comment (frozenhelium) [reply src/stories/Table/Table.stories.tsx:121]: Done!

File: src/stories/Table/Table.stories.tsx (hunk 2 above)
Comment (frozenhelium) [reply src/stories/Table/Table.stories.tsx:105]: Done!

File: src/index.tsx (hunk 3 above)
Comment (frozenhelium) [reply src/index.tsx]: Done!

File: src/components/Table/predefinedColumns.ts (hunk 5 above)
Comment (frozenhelium) [reply src/components/Table/predefinedColumns.ts]: Done!

File: src/stories/Table/Table.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    const orderState = useOrderState(staticColumnOrdering);
+    const { ordering } = orderState;
+
+    const orderedColumns = useOrdering(columns, ordering);
+    const filteredData = useFiltering(filtering, orderedColumns, data);
+    const sortedData = useSorting(sorting, orderedColumns, filteredData);
+     */
+
+    /*
```
Comment (tnagorra) [inline src/stories/Table/Table.stories.tsx:124]:
```suggestion

```

File: src/stories/Table/Table.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+import { Story } from '@storybook/react/types-6-0';
+import Table, {
+    Props as TableProps,
+    Column,
+} from '#components/Table';
+import useRowExpansion from '#components/Table/useRowExpansion';
+// import { FilterType } from '#components/Table/types';
+// import useFiltering, { useFilterState, FilterContext } from '#components/Table/useFiltering';
+// import useOrdering, { useOrderState, OrderContext } from '#components/Table/useOrdering';
+// import useSorting, { useSortState, SortContext } from '#components/Table/useSorting';
```
Comment (tnagorra) [inline src/stories/Table/Table.stories.tsx]: Please use block comment

File: src/stories/Table/Table.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+import {
+    createStringColumn,
+    createNumberColumn,
+    // createDateColumn,
+    // createDateTimeColumn,
+    // createYesNoColumn,
```
Comment (tnagorra) [inline src/stories/Table/Table.stories.tsx]: Please use block comment

File: src/components/Table/predefinedColumns.ts [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+            format: 'date',
+        }),
+        valueSelector: accessor,
+        valueComparator: (foo: D, bar: D) => compareDate(accessor(foo), accessor(bar)),
+    };
+    return item;
+}
+*/
+
+/*
```
Comment (tnagorra) [inline src/components/Table/predefinedColumns.ts]:
```suggestion

```

File: src/components/Table/useRowExpansion.tsx [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        const isActive = rowKey === expandedRow;
+
+        return (
+            <>
+                {React.cloneElement(row, {
+                    onClick: () => {
+                        setExpandedRow((oldValue) => (
+                            oldValue === rowKey ? undefined : rowKey
+                        ));
+                    },
```
Comment (tnagorra) [inline src/components/Table/useRowExpansion.tsx:47]: Please memoize this!

File: src/components/Table/useRowExpansion.tsx [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    options: ExpansionOptions = {},
+) {
+    const {
+        expandedRowClassName,
+        expandedCellClassName,
+        expansionCellClassName,
+        expansionRowClassName,
+    } = options;
+
+    const [expandedRow, setExpandedRow] = React.useState<K>();
```
Comment (tnagorra) [inline src/components/Table/useRowExpansion.tsx]: Does this work? 
```suggestion
    const [expandedRow, setExpandedRow] = React.useState<K | undefined>();
```

File: src/components/Table/useRowExpansion.tsx [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    const rowModifier: (
+        o: RowOptions<D, K>
+    ) => React.ReactElement = React.useCallback((rowOptions: RowOptions<D, K>) => {
```
Comment (tnagorra) [inline src/components/Table/useRowExpansion.tsx]:
```suggestion
    ) => React.ReactElement = React.useCallback((rowOptions) => {
```

File: src/hooks/useDropHandler.tsx (hunk 4 above)
Comment (frozenhelium) [reply src/hooks/useDropHandler.tsx:7]: Should not be related to this PR

File: src/components/Table/HeaderCell/index.tsx [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+    useContext,
+} from 'react';
+import { _cs } from '@togglecorp/fujs';
+
+import {
+    FaSortUp,
+    FaSortDown,
+    FaSort,
+    // FaLessThanEqual,
+    // FaGreaterThanEqual,
```
Comment (subinasr) [inline src/components/Table/HeaderCell/index.tsx]: Can we remove this if it's not being used?

File: src/components/Table/HeaderCell/index.tsx [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+            if (onFilterValueChange) {
+                onFilterValueChange(
+                    name,
+                    { ...filterValue, lessThanOrEqualTo: value },
+                );
+            }
+        },
+        [name, filterValue, onFilterValueChange],
+    );
+     */
```
Comment (subinasr) [inline src/components/Table/HeaderCell/index.tsx]: Same here

File: src/stories/Table/Table.stories.tsx (hunk 6 above)
Comment (subinasr) [reply src/stories/Table/Table.stories.tsx:124]: What!? I suppose the initial lines and the suggested change are supposed to be different?

**Review by subinasr — COMMENTED** [review]
Waiting for changes

File: src/hooks/useDropHandler.tsx (hunk 4 above)
Comment (tnagorra) [reply src/hooks/useDropHandler.tsx:7]: Yes, but please use

File: src/stories/Table/Table.stories.tsx (hunk 6 above)
Comment (tnagorra) [reply src/stories/Table/Table.stories.tsx:124]: They are. Red means deleted. Green means added.

Comment (subinasr) [reply src/stories/Table/Table.stories.tsx:124]: Lol. Okay

File: src/hooks/useDropHandler.tsx (hunk 4 above)
Comment (frozenhelium) [reply src/hooks/useDropHandler.tsx:7]: Okay

File: src/components/Table/useRowExpansion.tsx (hunk 10 above)
Comment (frozenhelium) [reply src/components/Table/useRowExpansion.tsx:47]: Cannot be done :(

File: src/hooks/useDropHandler.tsx (hunk 4 above)
Comment (frozenhelium) [reply src/hooks/useDropHandler.tsx:7]: Both have slightly different feature set, so it can't be done at the moment

**Pull request merged**